client/landing/stepper/declarative-flow/internals/hooks/use-preload-steps.ts

```diff
@@ -0,0 +1,79 @@
+import { SiteDetails } from '@automattic/data-stores';
+import debugFactory from 'debug';
+import { useEffect } from 'react';
+import { useSelector } from 'calypso/state';
+import { isUserLoggedIn } from 'calypso/state/current-user/selectors';
+import type { Flow, StepperStep } from '../types';
+
+const debug = debugFactory( 'calypso:stepper:preloading' );
+
+async function tryPreload( step?: StepperStep, followingStep?: StepperStep ) {
+	if ( step && 'asyncComponent' in step ) {
+		debug( 'Preloading next step:', step.slug );
+
+		await step.asyncComponent();
+
+		// Flows are indeterminate, they often pick one of the two next steps based on user input, so load two steps ahead.
+		tryPreload( followingStep );
+	}
+}
+
+/**
+ * Preload the next step in the flow, if it's an async component.
+ *
+ * @param siteSlugOrId The site slug or ID.
+ * @param selectedSite The selected site.
+ * @param currentStepRoute The current step route.
+ * @param flowSteps The flow steps.
+ * @param flow The flow.
+ */
+export function usePreloadSteps(
+	siteSlugOrId: string | number,
+	selectedSite: SiteDetails | undefined | null,
+	currentStepRoute: string,
+	flowSteps: readonly StepperStep[],
+	flow: Flow
+) {
+	const isLoggedIn = useSelector( isUserLoggedIn );
+
+	useEffect( () => {
+		if ( siteSlugOrId && ! selectedSite ) {
+			// If this step depends on a selected site, only preload after we have the data.
+			// Otherwise, we're still waiting to render something meaningful, and we don't want to
+			// potentially slow that down by having the CPU busy initialising future steps.
+			return;
+		}
+		if ( currentStepRoute ) {
+			// The user step is a special case, as it's not part of the flow steps. It always comes in the end of the steps array.
+			if ( currentStepRoute === 'user' ) {
+				// Load the first steps that requires authentication.
+				const nextStepIndex = flowSteps.findIndex( ( step ) => step.requiresLoggedInUser );
+				const nextStep = flowSteps[ nextStepIndex ];
+				const nextNextStep = flowSteps[ nextStepIndex + 1 ];
+				tryPreload( nextStep, nextNextStep );
+			} else {
+				// If any step requires authentication, preload the user step.
+				if ( ! isLoggedIn && flowSteps.some( ( step ) => step.requiresLoggedInUser ) ) {
+					const userStep = flowSteps.find( ( step ) => step.slug === 'user' );
+					tryPreload( userStep );
+				}
+
+				const nextStepIndex = flowSteps.findIndex( ( step ) => step.slug === currentStepRoute ) + 1;
+				const nextNextStepIndex =
+					flowSteps.findIndex( ( step ) => step.slug === currentStepRoute ) + 2;
```

Comment:
Rather than doing two searches, can we store the result in a const?
```suggestion
				const currentStepIndex = flowSteps.findIndex( ( step ) => step.slug === currentStepRoute );
				const nextStepIndex =  currentStepIndex + 1;
				const nextNextStepIndex = currentStepIndex + 2;
```

Reply:
Addressed in [`bc63d36` (#99352)](https://github.com/Automattic/wp-calypso/pull/99352/commits/bc63d36202b872be153a374ada7a5e4a94172d84)